app/assets/javascripts/angular/services/annotation_refresher.js.erb

```diff
@@ -1,31 +1,67 @@
-angular.module("Prometheus.services").factory('AnnotationRefresher', ["$http", "VariableInterpolator", function($http, VariableInterpolator) {
-  return function(graph, scope) {
-    var tags = graph.tags.map(function(e) {
-      if (e.name) {
-        var n = VariableInterpolator(e.name, scope.vars);
-        return n.split(",").map(function(s) { return s.trim(); });
-      } else {
-        return "";
-      }
-    });
-    if (tags.length) {
-      var range = Prometheus.Graph.parseDuration(graph.range);
-      var until = Math.floor(graph.endTime || Date.now()) / 1000;
+angular.module("Prometheus.services").factory('AnnotationRefresher', ["$http", "$q", "$timeout", "VariableInterpolator", function($http, $q, $timeout, VariableInterpolator) {
+  var MINUTE = 1000*60;
+  var promises = {};
+  var debounce;
+  return {
+    refresh: function(tags, range, endTime, vars) {
+      var tags = tags.map(function(e) {
+        if (e.name) {
+          var n = VariableInterpolator(e.name, vars);
+          return n.split(",").map(function(s) { return s.trim(); });
+        } else {
+          return "";
+        }
+      });
+
+      if (tags.length) {
+        var r = Prometheus.Graph.parseDuration(range);
+        var until = Math.floor(endTime || Date.now()) / 1000;
 
-      tags.forEach(function(t) {
-        $http.get('<%= Rails.configuration.path_prefix %>annotations', {
-          params: {
-            'tags[]': t,
-            until: until,
-            range: range
+        if (!endTime) {
+          // Set an endTime that can be removed from the promises map
+          // periodically.
+          endTime = "now";
+        }
+        var k = "" + range + endTime;
+        if (!promises[k]) {
+
+          if (endTime === "now") {
+            // If endTime isn't set, a long-running dashboard with
+            // auto-refresh will continually use annotations from the initial
+            // page load. This key needs to be deleted periodically.
+            // TODO: 5 min is arbitrary. Should it be something else?
+            var clear = function(key) {
+              return function() { promises[key] = undefined; };
+            }(k);
+            $timeout(clear, 5*MINUTE);
           }
-        })
-        .then(function(payload) {
-          scope.$broadcast('annotateGraph', payload.data.posts);
-        }, function(response) {
-          scope.errorMessages.push("Error " + response.status + ": Error occurred fetching annotations.");
-        });
-      });
+
+          promises[k] = tags.map(function(t) {
+            return $http.get('<%= Rails.configuration.path_prefix %>annotations', {
```

Comment:
I'm assuming we can just parse the templates in [`karma.rake`](https://github.com/prometheus/promdash/blob/master/lib/tasks/karma.rake), but that seems like making more work for ourselves instead of just removing the rails templating from our js files